### DISCUSSION TITLE
how to update page through a background task?

### DISCUSSION THREAD

**Discussion by zillionare** [h2oai/wave#1476] [Q&A] [answered]
consider this scenario:

user triggers an action on the UI, after very initial handling, a background task is started, and this task will emit some signals, I want wave app to receive the signals (with data), and update the page accordingly. You can think it's like trigger a training and report training loss to end user at every epoch's end.

How should I achieve this in wave?

ps, the possible way I can think of:
1. at the initial handler, bind the event listener with `q` objects, then it can use this obj to update page. Not sure how long will this object remain valid
2. at the background task, grab the `page` object and then go to the card, do update. However, my current route is '/#bt' ， seems it's not possible to get the reference.

**Comment by zillionare** [comment]
at second thought, the second option is invalid actually. I have to grab object `q` as this page belongs to specific user and should not be view by others, other browser tabs.

**Comment by mturoci** [comment]
Try looking at one of our examples:
https://wave.h2o.ai/docs/examples/background
https://wave.h2o.ai/docs/examples/background-executor
https://wave.h2o.ai/docs/examples/background-progress/ - this shows how to update UI, but first 2 are good to look through too.

**Comment by zillionare** [comment, marked as answer]
thanks @mturoci 

My case is a little bit complicated, due to the background task update progress through callbacks, so in a callback, I have to consider if the Q object still valid.

Finally it's achieved in this way, share it here maybe others may confront same issue:

upon launch background task, an asyncio.Event is created and passed to the callback, and let the main loop wait on this event. Once the callback get update, signal the event with received data so the main loop can update the page.

the main loop, will holds the Q object all the time.

background-executor maybe an neat way too, but sometimes it's hard to serialize objects through processes, especially when u don't know which objects are serialized and pass through, so most time I use pre-generated process pool (by subprocess), then communicate inter-process by message.

**Reply by mturoci** [reply]
Sounds good. Would you mind creating a PR with a minimal example? Might help others.
